ansible/eda-server
# feat: implement RBAC

# Description

This PR introduces 

1. A new RBAC functionality for EDA based on [RBAC library](https://github.com/ansible/django-ansible-base/tree/devel/ansible_base/rbac) from django-ansible-base.
2. Organization and teams models/endpoints.
3. Optimization for integration tests using general fixtures.

JIRA Epic: [AAP-19859](https://issues.redhat.com/browse/AAP-19859)

UI PR: https://github.com/ansible/ansible-ui/pull/2036 & https://github.com/ansible/ansible-ui/pull/2071
QE PR: https://github.com/ansible/eda-qa/pull/313

NOTE: we'll cut a new release in DAB once all fixes for RBAC are in and pin to that release version before merging this PR.

# Background

This RBAC implementation introduces a new way of managing roles and permissions that mainly consists of organization level and object level access. The existing 6 managed roles are migrated to org-level roles and apply to all resources in that organization. In addition to these roles, we are also pre-creating object-level managed roles that will grant object-level access to a user or a team. Creation of custom roles is also introduced that will allow admin users to create new custom roles with a list of permissions of their choice as per their needs.

Since some resources in EDA have parent-child relation, the managed object-level roles will also reflect this relation. For example, a project is a parent model for rulebooks thus assigning a user `Project Admin` role will also grant them full access to the project's rulebook objects. `Project Admin` role looks something like this:
```
{
      "id": 13,
      "name": "Project Admin",
      "description": "Has all permissions to a single project and its child resources - rulebook",
      "content_type": "eda.project",
      "permissions": [
        "eda.change_project",
        "eda.delete_project",
        "eda.view_project",
        "eda.view_rulebook"
      ],
}
```
These parent-child relations are defined [here](https://github.com/ansible/eda-server/blob/26f9ab792496325e7a53985b70b8837c42c8d6e3/src/aap_eda/core/models/__init__.py#L77).

Full documentation for RBAC library from DAB can be found [here](https://github.com/ansible/django-ansible-base/blob/devel/docs/apps/rbac.md). 

## URLs

The following urls are imported from DAB RBAC library:

1. `/api/eda/v1/role_definitions/` - used to create/retrieve/list roles that has a list of permissions on different resources, which will replace the current `/roles` endpoints. The POST payload looks as follows:
```
{
    "permissions": ["view_project"],
    "content_type": "eda.project",
    "name": "Project Viewer",
    "description": "custom role"
}
```
2. `/api/eda/v1/role_user_assignments/` - used to assign an existing role to a user. The POST endpoint payload looks as follows
```
{
    "role_definition": 1,
    "object_id": 1,
    "user": 1
}
```
`role_definition` is the role that is to be assigned to the user. `object_id` is the ID of the object the role applies to, and `user` is the ID of the user that the role is being assigned to. 
4. `/api/eda/v1/role_team_assignments/` - used to assign an existing role to a team. This is very similar to `role_user_assignment` but instead of a user the team gets the role assigned and all the members of the team inherit the role as well as its permissions. Example POST payload
```
{
    "role_definition": 3,
    "object_id": 3,
    "team": 2
}
```

In addition, we also implemented custom `api/eda/v1/organizations/` and `api/eda/v1/teams/` endpoints that are used for organization and team operations accordingly.

NOTE: Organization and Team memberships are also managed by `role_user_assignment` using managed roles called `Organization Member` and `Team Member` that will have the following permissions, `["eda.member_organization", "eda.view_organization"]` and `["eda.member_team", "eda.view_team"]`, respectively.

## Data Migration

1. The currently existing 6 system roles in EDA will be migrated to organization level roles. In addition to this, we are also creating object-level roles per resource in `create_initial_data.py`, e.g. Activation Admin, Project Admin, etc., that can be used for object level permissions.
2. All the existing resources data will be moved to a Default Organization. 


# Testing

Follow the steps below to test an object-level access for resources, such as Project, Decision Environment and Activation.
1. Log in as admin user. The user with `is_superuser=True` flag still works and has system level access.
2. Create a new Project and a new Decision Environment, and start up a new Activation.
3. Create a new user for testing through the API (the UI might expect roles value, thus we need to use the API for now). The payload can be as simple as this:
```
{
  "username": "doston",
  "password": "doston"
}
```
4. Go to `/api/eda/v1/role_definitions/` and get the ID for `Activation Admin`, which we will use in next steps.
5. Make a POST request to `/api/eda/v1/role_user_assignments/` with the following payload; make sure to replace `role_definition` value with the ID from step 4. `object_id` is the ID of the Activation we created in step 2 and user is the test user `doston` created in step 3. (we can ignore `object_ansible_id` and `user_ansible_id` for now)
```
{
  "object_id": "1",
  "role_definition": 14,
  "user": 3
}
```
This should give the user `doston` full access to the Activation we created earlier as well as its child objects such as Activation Instances, Audit Rules, Audit Events and Audit Actions.
6. Log out and log in as the user `doston` and check to make sure the user has access to the Activation in question. 

More roles can be assigned to the user to have access to other objects like Project and Decision Environments by creating new role user assignments.
